ecosystem/typescript/sdk_v2/src/types/account_address.ts

```diff
@@ -0,0 +1,241 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+import { bytesToHex, hexToBytes } from "@noble/hashes/utils";
+import { Serializer, Deserializer, Bytes } from "../bcs";
+
+// TODO: Get feedback on these error messages, ideally they can be used directly in
+// dapps and wallets. The enum is meant to make it possible for daps to user their
+// own error messages if they want, but the defaults should still be good, and they're
+// not great right now.
+
+// These options influence fromStr and the isValid* functions. You may use them to make
+// the parsing of those functions strictier. Not setting these options is the best way
+// to comply with AIP 40 but while migrating it may be helpful to be able to enforce
+// stricter behavior. If we agree to make these two enabled by default, I can update
+// the AIP to say that this is what people SHOULD do but they MAY opt in to accepting
+// addresses more leniently in specifically these two ways.
+export type ValidityOptions = {
+  // Only addresses in the long form (64 characters + optionally a leading 0x) will be
+  // considered valid unless the address is special.
+  requireLongFormUnlessSpecial?: boolean;
+
+  // Only addresses in the long form (64 characters + optionally a leading 0x) will be
+  // considered valid, even if the address is special.
+  requireLongForm?: boolean;
+
+  // If set, the address string will be considered invalid if there is no leading 0x.
+  requireLeadingZeroX?: boolean;
+};
+
+// TODO: As part of reviewing this we should discuss these options. I'm partially
+// inclined to say that requireLongFormUnlessSpecial and requireLeadingZeroX should
+// actually be enabled by default. Maayan and Greg to provide their thoughts as well as
+// wallet and dapp folks.
+export const defaultValidityOptions: ValidityOptions = {
+  requireLongFormUnlessSpecial: false,
+  requireLongForm: false,
+  requireLeadingZeroX: false,
+};
+
+export enum AddressInvalidReason {
+  INCORRECT_NUMBER_OF_BYTES = "incorrect_number of bytes",
+  ADDRESS_TOO_SHORT = "address_too_short",
+  ADDRESS_TOO_LONG = "address_too_long",
+  LEADING_ZERO_X_REQUIRED = "leading_zero_x_required",
+  LONG_FORM_REQUIRED = "long_form_required",
+  LONG_FORM_REQUIRED_UNLESS_SPECIAL = "long_form_required_unless_special",
+}
+
+/**
+ * This error is used to explain why an address was invalid.
+ */
+export class AddressInvalidError extends Error {
```

Comment:
I was thinking on how the SDK should handle errors. One option is to throw an error as we do (slightly) in v1. I found it a bit annoying as the Error class is a bit weird in the way that by default it handles and logs only the `message` part. For example, with the `AddressInvalidError`, the message will be logged but the `reason` part not. I was thinking on maybe stop throwing errors and instead start returning Object with a `status`, `message`, etc. something like
```
type AptosResponse = {
 status: "success" | "error"
 message: string
}
```

also wonder what you all think @xbtmatt @gregnazario @0xjinn

Reply:
Only problem with returning a response that's not an error, is that now you can't easily use `try catch` and instead have to change your whole flow to use type matching on errors.  I'm not sure how that works with the average developer

Reply:
I agree JS errors are sorta nasty for exactly the reason you describe Maayan, but I also agree with Greg that returning essentially a Result is not really standard for JS.